Support Uuid/Ulid as user identifier in hasUserChanged()

spackmat opened the pull request:
Update: Revoked this PR, because implementing `UserInterface::getUserIdentifier()` the right way by returning a string instead of the `Uid`-object makes its change unnecessary.

| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | no (completes a new feature of Symfony 5.2)
| Deprecations? | no
| Tickets       | none
| License       | MIT
| Doc PR        | 

This was my problem: Uuid/Ulid are officially supported as Doctrine types since Symfony 5.2, but when a User Entity uses them as its identifier, the strict comparison of the identifier in `AbstractToken::hasUserChanged()` fails unless the User Entity implements `EquatableInterface` and does the whole comparison itself. In effect, the user is always assumed as changed and logged out. This little pitfall was driving me mad, when I implemented my first Uid-identified User entity. I was so close and read the documentation over and over again, until I found out that the comparison of my identifier went wrong.

This change had implemented a small check, if the `getUserIdentifier()` method of the User Entity returns an instance of a `AbstractUid` object and uses its equals-method for the comparison instead of the strict `!==` operator, that is always true for two different Uuid/Ulid objects, even if they contain the same Uid. Since the `UserInterface::getUserIdentifier()` method should return a string, this is not necessary at all.

spackmat committed:
Support Uuid/Ulid as user identifier in hasUserChanged()

Uuid/Ulid are officially supported as Doctrine types since Symfony 5.2, but when a User Entity uses them as its identifier, the strict comparison of the identifier in `AbstractToken::hasUserChanged()` fails unless the User Entity implements `EquatableInterface` and does the whole comparison itself. In effect, the user is always assumed as changed and logged out. This little pitfall is driving every Dev mad, who implements his/her first Uid-identified User entity. One is so close and reads the documentation over and over again, until this little reason is found.

This change implements a small check, if the `getUserIdentifier()` method of the User Entity returns an object with an equals method and uses that for the comparison instead of the strict `!==` operator, that is always true for two different Uuid/Ulid objects, even if they contain the same Uid.

spackmat committed:
Meet coding standards.

Two small changes to meet the coding standard test at fabpot.io

spackmat committed:
Meet coding standards the second time

Should read the whole report.

derrabus commented:
Thank you for your PR. I agree that this could be handled better.

Please target the 5.4 branch. I don't think that this change qualifies for a bugfix. And please add a test that covers your change.

derrabus commented on src/Symfony/Component/Security/Core/Authentication/Token/AbstractToken.php:
```diff
@@ -309,7 +309,12 @@ private function hasUserChanged(UserInterface $user): bool
         };
-        if ($userIdentifier($this->user) !== $userIdentifier($user)) {
+        // if the username/identifier is an object with an eqials-method (like Uuid or Ulid), prefer that for a comparison
+        if (\is_object($userIdentifier($this->user)) && method_exists($userIdentifier($this->user), 'equals')) {
```
I don't feel comfortable with using duck typing here. Maybe we can check for `instanceof AbstractUid` instead?

nicolas-grekas replied:
what about an even more generic interface as a contract?
`EquatableInterface::equals(mixed): bool`

spackmat replied:
I did think about that, but then decided not to implement a lex Uid, but a more abstract solution that solves the problem for all object-based identifiers that have an equals method. Maybe someone uses ShortID or other implementations of Uid.

But you are right, that duck typing opens the door widely for nasty other bugs, so maybe it 's not worth it and a lex Uid is a more stable solution. I'll change that.

spackmat committed:
Don't duck type the identifier object

but implement a lex Uid by checking for `instanceof \Symfony\Component\Uid\AbstractUid` instead

spackmat commented:
I removed the duck typing and changed the branch to 5.4. I can build a test case for that later, for now fighting with that problem pushed me closer to a tight deadline and I have to get that stuff ready before.

spackmat commented:
While writing the test, I found out that `UserInterface::getUserIdentifier()` must return a string from 6.0 onwards, so my patch here is only necessary for the 5.4 branch, when the User Entity is implemented in a way that ist not really compatible to the `UserInterface`. When I return the Uuid as a string in `MyUuidBasedUser::getUserIdentifier()`, the whole thing works fine. Didn't see that, because the method signature for `UserInterface::getUserIdentifier()` is only mentioned as a comment in the 5.3 and 5.4 branch.

So that was my fault and everything works fine with Uuid/Ulid based User Entities.

derrabus commented:
Makes sense. Anyway, thanks for proposing!
